Issue 1746 traits

pondermatic opened the pull request:
## Description
1. Added 3 traits.
    1. `LLMS_Trait_Audio_Video_Embed`
    2. `LLMS_Trait_Sales_Page`
    3. `LLMS_Trait_Singleton`.
1. Updated `LLMS_Loader->autoload()` to handle traits.
1. Moved the loading of the `LLMS_Loader` class from the `LifterLMS` class to the main `lifterlms.php` file.
1. Added unit tests for the new traits.
1. Added traits to the coding standards document.

Fixes #1746

## How has this been tested?
* Used Xdebug to step through the code.
* Ran the full suite of unit tests.
* Monitored the PHP error log.

## Types of changes
What types of changes does your code introduce?  Structural.

## Checklist:
- [x] My code has been tested.
- [x] My code passes all existing automated tests. <!-- Check code: `composer run-script tests-run`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/tests/README.md -->
- [x] My code follows the LifterLMS Coding & Documentation Standards. <!-- Check code: `composer run-script check-cs-errors`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/docs/coding-standards.md and https://github.com/gocodebox/lifterlms/blob/trunk/docs/documentation-standards.md -->

pondermatic committed:
Add traits to `autoload()`.

pondermatic committed:
Move sales page methods and properties to a new trait.

pondermatic committed:
Replace singleton code with `LLMS_Trait_Singleton`.

pondermatic committed:
Refactor audio and video code to `LLMS_Trait_Audio_Video_Embed`.

pondermatic committed:
Require that the exhibiting class implements the `add_properties()` and `get()` methods.
Document the requirement of the `LLMS_Post_Model::$model_post_type` property.
Improve filter DocBlocks.

pondermatic committed:
Add sales page trait tests.

pondermatic committed:
Add singleton trait tests.

pondermatic committed:
Add audio video embed trait tests.

pondermatic committed:
Link to trait file in test class DocBlock.

pondermatic committed:
Add traits to the coding standards doc.

pondermatic committed:
Remove redundant "with".

thomasplevy commented on class-lifterlms.php:
```diff
@@ -23,8 +23,13 @@
+ * @method static LifterLMS instance()
  */
-final class LifterLMS {
+final class LifterLMS  {
```
(Remove extra space)

```suggestion
final class LifterLMS {
```

thomasplevy commented on includes/traits/llms-trait-singleton.php:
```diff
@@ -0,0 +1,42 @@
+ *
+ * @package LifterLMS/Traits
+ *
+ * @since   [version]
```
```suggestion
 * @since [version]
```

thomasplevy commented on lifterlms.php:
```diff
@@ -47,6 +44,10 @@
 	define( 'LLMS_PLUGIN_DIR', dirname( __FILE__ ) . '/' );
 }
 
+// Autoloader.
```
File needs the `@version` tag updated

pondermatic replied:
The @version tag of line 8? OK.

Does the version on line 13 need changed here, or does the build process change it?

thomasplevy commented on includes/class-llms-loader.php:
```diff
@@ -14,6 +14,7 @@
  * LLMS_Loader
  *
  * @since 4.0.0
+ * @since [version] Add traits to `autoload()`.
```
File needs version update also

thomasplevy commented on includes/traits/llms-trait-singleton.php:
```diff
@@ -0,0 +1,42 @@
+	public static function instance() {
+
+		if ( is_null( self::$instance ) ) {
+			self::$instance = new self();
```
Since most of the singletons use `_instance` as a protected variable to hold the instance we should cautiously deprecate the use of this variable.

If we *retain* the existence of the `$_instance` variable on all classes that this has been removed from we can run a `property_exists()` check against the property, if it does, we should set it alongside the replacement `$instance`.

We also need to note the deprecations, retaining the variable on the classes gives us a place to note the deprecation with an `@deprecated` tag.

Alternatively I supposed a magic `__get()` can be used instead but then we get into inheritance issues with classes that already use a `__get()` method...

pondermatic replied:
I like the idea of keeping `$_instance`, marking it as deprecated, and setting it in `LLMS_Trait_Singleton::instance()` if it exists.

thomasplevy commented on includes/class-llms-events.php:
```diff
@@ -19,15 +19,13 @@
  */
 class LLMS_Events {
 
-	/**
```
should be retained and marked as deprecated

thomasplevy commented on includes/class-llms-grades.php:
```diff
@@ -14,15 +14,13 @@
-	 *
-	 * @var  null
-	 */
-	protected static $_instance = null;
```
should be retained and marked as deprecated

thomasplevy commented on includes/class.llms.achievements.php:
```diff
@@ -30,14 +35,6 @@ class LLMS_Achievements {
 
 	private $_content_type;
 
-	protected static $_instance = null;
```
should be retained and marked as deprecated

thomasplevy commented on includes/class.llms.certificates.php:
```diff
@@ -24,15 +24,13 @@
-	 *
-	 * @var LLMS_Certificates
-	 */
-	protected static $_instance = null;
```
should be retained and marked as deprecated

thomasplevy commented on includes/class.llms.emails.php:
```diff
@@ -17,37 +17,21 @@
-	 *
-	 * @var LLMS_Emails
-	 */
-	protected static $_instance = null;
```
should be retained and marked as deprecated

thomasplevy commented on includes/class.llms.engagements.php:
```diff
@@ -29,25 +34,6 @@ class LLMS_Engagements {
-	 *
-	 * @var LLMS_Engagements
-	 */
-	protected static $_instance = null;
```
should be retained and marked as deprecated

thomasplevy commented on includes/class.llms.payment.gateways.php:
```diff
@@ -15,38 +15,21 @@
-	 *
-	 * @var null
-	 */
-	private static $_instance = null;
```
should be retained and marked as deprecated

thomasplevy commented on includes/class.llms.integrations.php:
```diff
@@ -17,10 +17,13 @@
  */
 class LLMS_Integrations {
 
-	protected static $_instance = null;
```
should be retained and marked as deprecated

thomasplevy commented on includes/processors/class.llms.processors.php:
```diff
@@ -40,27 +45,6 @@ class LLMS_Processors {
-	 *
-	 * @var null
-	 */
-	protected static $_instance = null;
```
should be retained and marked as deprecated

thomasplevy commented on includes/notifications/class.llms.notifications.php:
```diff
@@ -19,15 +19,13 @@
-	 *
-	 * @var LLMS_Notifications
-	 */
-	protected static $_instance = null;
```
should be retained and marked as deprecated

thomasplevy commented on includes/models/model.llms.course.php:
```diff
@@ -16,9 +16,10 @@
- * @since [version] Check for an empty sales page URL or ID.
+ * @since [version] Add `__construct()`.
+ *              Move audio and video embed properties and methods to `LLMS_Trait_Audio_Video_Embed`.
+ *              Move sales page properties and methods to `LLMS_Trait_Sales_Page`.
```
New method addition is unnecessary in the class docs (the addition of the method can be seen on the method's changelog inherently)

```suggestion
 * @since [version] Move audio and video embed properties and methods to `LLMS_Trait_Audio_Video_Embed`.
 *              Move sales page properties and methods to `LLMS_Trait_Sales_Page`.
```

thomasplevy commented on includes/models/model.llms.lesson.php:
```diff
@@ -20,8 +20,9 @@
  * @since 4.4.0 Improve the query used to retrieve the previous/next so that we don't miss sibling lessons within the same section
  *              if the previous/next one(s) status is (are) not published. Make sure to always return `false` if no previous lesson is found.
  *              Use strict comparisons where needed.
+ * @since [version] Add `__construct()`.
+ *              Move audio and video embed properties and methods to `LLMS_Trait_Audio_Video_Embed`.
```
```suggestion
 * @since [version] Move audio and video embed properties and methods to `LLMS_Trait_Audio_Video_Embed`.
```

pondermatic replied:
Remove "Add `__construct()`." from the class changelog?

thomasplevy commented on includes/traits/llms-trait-audio-video-embed.php:
```diff
@@ -0,0 +1,117 @@
+ * **Classes that use this trait must call {@see LLMS_Trait_Audio_Video_Embed::construct_audio_video_embed}
+ * in their constructor.**
+ *
+ * @since [version] Refactored from `LLMS_Course`, `LLMS_Lesson` and `LLMS_Post_Model`.
```
```suggestion
 * @since [version]
```

thomasplevy commented on includes/traits/llms-trait-audio-video-embed.php:
```diff
@@ -0,0 +1,117 @@
+ *
+ * @package LifterLMS/Traits
+ *
+ * @since   [version]
```
```suggestion
 * @since [version]
```

thomasplevy commented on includes/traits/llms-trait-audio-video-embed.php:
```diff
@@ -0,0 +1,117 @@
+	public function get_video() {
+		return $this->get_embed( 'video' );
+	}
+}
```
Should be a newline at the end of the file: https://github.com/gocodebox/lifterlms/blob/8c913be8492b8216f8657ae2864ac0411c8d8700/.editorconfig#L15

thomasplevy commented on includes/traits/llms-trait-sales-page.php:
```diff
@@ -0,0 +1,138 @@
+ *
+ * @package LifterLMS/Traits
+ *
+ * @since   [version]
```
```suggestion
 * @since [version]
```

thomasplevy reviewed (changes requested):
Overall this is very good, I have quite a few notes, mostly around docs and few additional items:

Additional notes:

+ most of the file headers need `@version` updated
+ Interfaces should be left in the codebase and marked as deprecated
+ LitferLMS Assignments uses the audio and video interfaces on the `LLMS_Assignment` model (https://github.com/gocodebox/lifterlms-assignments/issues/63)

pondermatic committed:
Tweak code formatting.

pondermatic committed:
Tweak @since and @version DocBlock tags.

pondermatic committed:
Revert the removal of the `$_instance` property and mark it as deprecated.

pondermatic committed:
Update @version DocBlock tags.

pondermatic commented on tests/phpunit/unit-tests/traits/llms-test-trait-singleton.php:
```diff
@@ -0,0 +1,64 @@
+		$this->assertIsNotObject( $actual );
+
+		/** Test {@see LLMS_Trait_Singleton::instance()}. */
+		$instance1 = $this->mock->instance();
```
Use the scope Resolution operator `::` when calling the static `instance()` method.

pondermatic committed:
Handle deprecated `$_instance` property to prevent breaking 3rd party plugins, for now.

pondermatic committed:
Add testing of the deprecated $_instance property.
Replace the $mock property, that held an anonymous object, with a $mock_class property, that holds the class name of the anonymous object.

pondermatic reviewed (commented):
Fixed in commit #9b651ebe4e8dc286a86414ea8ff56e3fae7ca180.

pondermatic committed:
Remove note about adding `__construct()` in the class DocBlock.

pondermatic committed:
Remove note about where the methods and properties came from in the class DocBlock.

pondermatic committed:
Merge branch 'dev' into issue-1746-traits

pondermatic committed:
Revert the deletion of the interface files and mark them as deprecated.

pondermatic committed:
Merge branch 'issue-1746-traits' of https://github.com/pondermatic/lifterlms into issue-1746-traits

pondermatic committed:
Revert the additions of `@method static X instance()` from classes in favor of changing `@return` on `LLMS_Trait_Singleton::instance()` from `object` to `self`.
This keeps IDEs happy while reducing DocBlock size.

pondermatic commented:
There is a disturbance in the Force. Some unit tests are failing. Investigating now.

pondermatic committed:
Update for possible deprecated `$_instance` property.

pondermatic committed:
Use an actual WP post type to avoid `Unexpected incorrect usage notice for map_meta_cap` errors when running the full suite of unit tests.

thomasplevy commented on includes/class-llms-events.php:
```diff
@@ -19,13 +19,18 @@
-	 * Singleton instance
+	 * Singleton instance.
+	 *
+	 * @deprecated [version] Use {@see LLMS_Trait_Singleton::$instance}.
```
Direct reference to the replacement var will not be possible since it's a private class variable. So we should recommend reference to the function rather than the variable.

```suggestion
	 * @deprecated [version] Use {@see self::instance()}.
```

pondermatic replied:
Surprisingly, PhpStorm doesn't auto-link to `self::instance()`. How about this?
```suggestion
	 * @deprecated [version] Use {@see LLMS_Trait_Singleton::instance()}.
```

thomasplevy replied:
:+1:

thomasplevy commented on includes/class-llms-grades.php:
```diff
@@ -14,13 +14,18 @@
-	 * Singleton instance
+	 * Singleton instance.
+	 *
+	 * @deprecated [version] Use {@see LLMS_Trait_Singleton::$instance}.
```
Same

thomasplevy commented on includes/class-llms-sessions.php:
```diff
@@ -15,13 +15,18 @@
+	 * Singleton instance.
 	 *
-	 * @var null
+	 * @deprecated [version] Use {@see LLMS_Trait_Singleton::$instance}.
```
Same

thomasplevy commented on includes/class.llms.achievements.php:
```diff
@@ -30,14 +33,15 @@ class LLMS_Achievements {
+	/**
+	 * Singleton instance.
+	 *
+	 * @deprecated [version] Use {@see LLMS_Trait_Singleton::$instance}.
```
Same

thomasplevy commented on includes/class.llms.certificates.php:
```diff
@@ -24,11 +24,16 @@
-	 * Instance
+	 * Singleton instance.
+	 *
+	 * @deprecated [version] Use {@see LLMS_Trait_Singleton::$instance}.
```
Same

thomasplevy commented on includes/class.llms.emails.php:
```diff
@@ -17,36 +17,27 @@
-	 * protected private instance of email
+	 * Singleton instance.
+	 *
+	 * @deprecated [version] Use {@see LLMS_Trait_Singleton::$instance}.
```
Same

thomasplevy commented on includes/class.llms.engagements.php:
```diff
@@ -18,35 +18,28 @@
-	 * Protected instance of class
+	 * Singleton instance.
+	 *
+	 * @deprecated [version] Use {@see LLMS_Trait_Singleton::$instance}.
```
Same

thomasplevy commented on includes/class.llms.payment.gateways.php:
```diff
@@ -15,37 +15,27 @@
+	 * Singleton instance.
 	 *
-	 * @var null
+	 * @deprecated [version] Use {@see LLMS_Trait_Singleton::$instance}.
```
Same

eri-trabiccolo commented on tests/phpunit/unit-tests/class-llms-test-main-class.php:
```diff
@@ -27,15 +28,14 @@ public function setUp() {
 	 *
 	 * @since 3.3.1
-	 *
-	 * @return void
```
are we deciding to remove the return type from tests? Or only when it's void?

thomasplevy replied:
We haven't decided anything like this

thomasplevy replied:
Restore this line

pondermatic replied:
From the [WP PHP documentation standard](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#1-functions-class-methods):

> @return void should not be used outside of the default bundled themes.

Also, the [phpDocumentor](https://docs.phpdoc.org/3.0/guide/references/phpdoc/tags/return.html) site:

> functions and methods without a `return` value, the @return tag MAY be omitted here, in which case @return void is implied.

pondermatic committed:
Update @version DocBlock tags.

thomasplevy commented on includes/notifications/class.llms.notifications.php:
```diff
@@ -19,11 +19,16 @@
-	 * Singleton instance
+	 * Singleton instance.
+	 *
+	 * @deprecated [version] Use {@see LLMS_Trait_Singleton::$instance}.
```
Same

thomasplevy commented on includes/processors/class.llms.processors.php:
```diff
@@ -41,25 +44,13 @@ class LLMS_Processors {
+	 * Singleton instance.
 	 *
-	 * @since 3.15.0
+	 * @deprecated [version] Use {@see LLMS_Trait_Singleton::$instance}.
```
Same

pondermatic committed:
Remove the redundant @since changelog from the file header.

pondermatic committed:
Remove the @version tag from the class DocBlock.
@version tags are only used in file DocBlocks and unit tests don't have file headers.

thomasplevy commented on includes/traits/llms-trait-singleton.php:
```diff
@@ -0,0 +1,54 @@
+	 */
+	public static function instance() {
+
+		if ( is_null( self::$instance ) ) {
+
+			if ( property_exists( __CLASS__, '_instance') ) {
+
+				if ( is_null( self::$_instance ) ) {
+
+					self::$instance = new self();
+					self::$_instance = self::$instance;
+				} else {
+					self::$instance = self::$_instance;
+				}
+			} else {
+				self::$instance = new self();
+			}
+		}
```
I might be missing something but I don't think the the `is_null()` condition on line 40 is necessary. This should get the job done the same, right?

The idea being that `$_instance` should be set to be equal to `$instance` when the property exists. EG: these two var should always be either both `null` or both equal to the same instance so an additional `is_null()` on the `_instance` var shouldn't ever be necessary.

```suggestion
		if ( is_null( self::$instance ) ) {
			self::$instance = new self();
			if ( property_exists( __CLASS__, '_instance' ) ) {
				self::$_instance = self::$instance;
			}
		}
```

pondermatic replied:
If a 3rd party developer extends a LifterLMS class  with the `$_instance` property, **and** their code sets it, then we should use what is in `$_instance`. I know that this is a highly unlikely edge case.

thomasplevy replied:
You're right but I hate it. :+1:

pondermatic committed:
Change @deprecated description from the private LLMS_Trait_Singleton::$instance to the public LLMS_Trait_Singleton::instance().

pondermatic committed:
Revert the removal of `@return void`.

thomasplevy commented on tests/phpunit/unit-tests/traits/llms-test-trait-singleton.php:
```diff
@@ -0,0 +1,139 @@
+		parent::setUp();
+
+		# Instantiate an anonymous class that uses the trait to be tested.
+		$mock = new class {
+
+			use LLMS_Trait_Singleton;
+
+			protected $color;
+
+			protected static $_instance = null;
+
+			public static function deprecated_instance() {
+				if ( is_null( self::$_instance ) ) {
+					self::$_instance = new self();
+				}
+
+				return self::$_instance;
+			}
+
+			public static function init() {
+				self::$_instance = null;
+				self::$instance  = null;
+			}
+
+			public function get_color() {
+				return $this->color;
+			}
+
+			public function set_color( $color ) {
+				$this->color = $color;
+			}
+		};
+
+		$this->mock_class = get_class( $mock );
```
I think this will fix tests failing on php8

```suggestion
		$this->mock_class = new class {

			use LLMS_Trait_Singleton;

			protected $color;

			protected static $_instance = null;

			public static function deprecated_instance() {
				if ( is_null( self::$_instance ) ) {
					self::$_instance = new self();
				}

				return self::$_instance;
			}

			public static function init() {
				self::$_instance = null;
				self::$instance  = null;
			}

			public function get_color() {
				return $this->color;
			}

			public function set_color( $color ) {
				$this->color = $color;
			}
		};
```

pondermatic replied:
I originally wrote it this way, storing an object. But then I realized that it wasn't a true test of static properties.

I'm looking into what's up with PHP 8 and the Reflection classes.

pondermatic replied:
[`ReflectionProperty::getValue()`](https://www.php.net/manual/en/reflectionproperty.getvalue.php), used by [`LLMS_Unit_Test_Util ::get_private_property_value()`](https://github.com/gocodebox/lifterlms-tests/blob/trunk/framework/class-llms-unit-test-util.php#L93), requires an object for non-static properties and `null` for static properties.

Options:
1. Update `LLMS_Unit_Test_Util ::get_private_property_value()` to check if property is static.
2. Add a new method to handle private static properties. `LLMS_Unit_Test_Util ::get_private_static_property_value()` 
3. Use [`ReflectionClass::getDefaultProperties()`](https://www.php.net/manual/en/reflectionclass.getdefaultproperties.php) instead of `ReflectionProperty::getValue()`. **Needs testing.**

thomasplevy replied:
1. sounds good

pondermatic replied:
After some more reading, I agree that #1 is the best option.

pondermatic replied:
Of course, this means that now this PR is dependent on a lifterlms-tests PR.

thomasplevy replied:
That's okay, it happens

pondermatic replied:
I made a lifterlms-tests PR https://github.com/gocodebox/lifterlms-tests/pull/9.

thomasplevy commented:
Also need to fix CS issues: https://app.travis-ci.com/github/gocodebox/lifterlms/jobs/532956739#L313-L351

pondermatic committed:
Reformat to adhere to coding standards.
Thank you, PHP CodeSniffer!

thomasplevy commented:
Closing in favor of #1758 (which includes test lib updates)
